File: ts/project/.sdk/src/cmp/rust/fragment/EntityListOp.fragment.rs
// EJECT-START

fn list(self: &Rc<Self>, reqmatch: Value, ctrl: Value) -> Result<Vec<Rc<Self>>, ProjectNameError> {
    let ctx = self.utility.make_context(
        CtxSpec {
            opname: Some("list".to_string()),
            ctrl: Some(ctrl),
            mtch: Some(self.mtch.borrow().clone()),
            data: Some(self.data.borrow().clone()),
            reqmatch: Some(reqmatch),
            ..Default::default()
        },
        Some(&self.ent_ctx()),
    );

    let out = self.run_op(&ctx, &|ctx| {
        if let Some(result) = ctx.result.borrow().clone() {
            let resmatch = result.borrow().resmatch.clone();
            if let Value::Map(_) = resmatch {
                *self.mtch.borrow_mut() = resmatch;
            }
        }
    })?;

    // `list` resolves to one ENTITY per record. makeResult cannot build them
    // here — it works in `Value`, which is a closed data union with no slot
    // for an entity — so the op does, mirroring what the dynamic targets get
    // from makeResult. See AGENTS.md "Entity operations return ENTITIES".
    let mut items: Vec<Rc<Self>> = Vec::new();
    if let Value::List(list) = &out {
        for entry in list.borrow().iter() {
            let ent = EntyClass::new(&self.client, vs::clone(&self.entopts));
            if let Value::Map(_) = entry {
                ent.data(Some(entry));
            }
            items.push(ent);
        }
    }

    Ok(items)
}

// EJECT-END

File: ts/project/.sdk/src/cmp/zig/fragment/EntityListOp.fragment.zig
// EJECT-START

pub fn list(self: *EntyClass, reqmatch: Value, ctrl: Value) EntListResult {
    const ctx = self.utility.make_context(CtxSpec{
        .opname = "list",
        .ctrl = ctrl,
        .mtch = self.mtch,
        .data = self.data,
        .reqmatch = reqmatch,
    }, self.ent_ctx());
    const out = switch (self.run_op(ctx, list_post_done)) {
        .err => |e| return EntListResult{ .err = e },
        .ok => |v| v,
    };

    // `list` resolves to one ENTITY per record. make_result cannot build them
    // here — it works in Value, which has no slot for an entity — so the op
    // does, mirroring what the dynamic targets get from make_result.
    var items = std.ArrayList(*EntyClass).init(h.A());
    if (out == .array) {
        for (out.array.data.items) |entry| {
            const ent = EntyClass.new(self.client, h.clone(self.entopts));
            if (entry == .object) {
                _ = ent.data_impl(entry);
            }
            items.append(ent) catch {};
        }
    }

    return EntListResult{ .ok = items.toOwnedSlice() catch &[_]*EntyClass{} };
}

fn list_post_done(self: *EntyClass, ctx: *Context) void {
    if (ctx.result) |result| {
        const resmatch = result.resmatch;
        if (resmatch == .object) self.mtch = resmatch;
    }
}

// EJECT-END
